File: examples/python/reverse_proxy/reverse_proxy.py
#!/usr/bin/env python3
"""MockServer Python client -- reverse proxy example.

Configures MockServer as a reverse proxy that routes requests to different
backends based on the path:

  /rest/*   -> local backend (127.0.0.1:8080, HTTP)
  /*        -> QA environment (192.168.50.10:443, HTTPS)

Run this script against a MockServer on localhost:1080.
"""

from mockserver import (
    HttpForward,
    HttpRequest,
    MockServerClient,
    Times,
)


MOCK_HOST = "localhost"
MOCK_PORT = 1080


def configure_reverse_proxy(client: MockServerClient) -> None:
    """Set up two forwarding rules: REST API to local, everything else to QA."""

    # Rule 1: forward /rest/* to a local backend
    client.when(
        HttpRequest(path="/rest.*"),
        times=Times(unlimited=True),
    ).forward(
        HttpForward(host="127.0.0.1", port=8080, scheme="HTTP")
    )
    print("proxy rule created: /rest/* -> 127.0.0.1:8080 (HTTP)")

    # Rule 2: forward everything else to QA environment
    client.when(
        HttpRequest(path="/.*"),
        times=Times(unlimited=True),
    ).forward(
        HttpForward(host="192.168.50.10", port=443, scheme="HTTPS")
    )
    print("proxy rule created: /* -> 192.168.50.10:443 (HTTPS)")


def main() -> None:
    with MockServerClient(MOCK_HOST, MOCK_PORT) as client:
        client.reset()
        configure_reverse_proxy(client)
        print("\nReverse proxy configured. MockServer is now forwarding traffic.")
        print("Press Ctrl-C to exit (expectations remain until MockServer is reset).")

        # Keep the script alive so the user can observe the proxy in action.
        # In a real scenario you might not need this -- the expectations persist
        # on the server after the client disconnects.
        try:
            import time
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            print("\nExiting.")


if __name__ == "__main__":
    main()

File: examples/ruby/reverse_proxy/reverse_proxy.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Demonstrates using MockServer as a reverse proxy.
#
# This sets up MockServer to:
#   1. Forward REST API requests to a local backend (e.g. localhost:8080)
#   2. Forward all other requests to a remote environment
#
# Prerequisites: MockServer running on localhost:1080
#   docker run -d -p 1080:1080 mockserver/mockserver

require 'mockserver-client'

include MockServer

client = Client.new('localhost', 1080)

# ------------------------------------------------------------------
# 1. Forward REST API requests to a local backend
# ------------------------------------------------------------------
client.when(
  HttpRequest.request(path: '/rest.*'),
  times: Times.unlimited
).forward(
  HttpForward.new(host: '127.0.0.1', port: 8080, scheme: 'HTTP')
)
puts '1. Forwarding /rest.* -> 127.0.0.1:8080 (HTTP)'

# ------------------------------------------------------------------
# 2. Forward all other requests to a remote environment
# ------------------------------------------------------------------
client.when(
  HttpRequest.request(path: '/.*'),
  times: Times.unlimited
).forward(
  HttpForward.new(host: '192.168.50.10', port: 443, scheme: 'HTTPS')
)
puts '2. Forwarding /.* -> 192.168.50.10:443 (HTTPS)'

puts "\nReverse proxy configured."
puts 'REST API requests go to the local backend; everything else goes to the remote environment.'
puts 'Press Ctrl-C to stop.'

# ------------------------------------------------------------------
# Wait for Ctrl-C
# ------------------------------------------------------------------
begin
  sleep
rescue Interrupt
  puts "\nShutting down..."
  client.reset
  puts 'All expectations cleared.'
end
